src/python/modules/PyTorch/gmm_objective.py

```diff
@@ -20,52 +28,54 @@ def sqsum(x):
 def log_wishart_prior(p, wishart_gamma, wishart_m, sum_qs, Qdiags, icf):
     n = p + wishart_m + 1
     k = icf.shape[0]
-    out = 0
-    for ik in range(k):
-        frobenius = sqsum(Qdiags[ik, :]) + sqsum(icf[ik, p:])
-        out = out + 0.5 * wishart_gamma * wishart_gamma * \
-            frobenius - wishart_m * sum_qs[ik]
-    C = n * p * (math.log(wishart_gamma) - 0.5 * math.log(2)) - log_gamma_distrib(0.5 * n, p)
-    return out - k * C
+
+    out = torch.sum(
+        0.5 * wishart_gamma * wishart_gamma *
+        (torch.sum(Qdiags ** 2, dim = 1) + torch.sum(icf[:,p:] ** 2, dim = 1)) -
+        wishart_m * sum_qs
+    )
+
+    C = n * p * (math.log(wishart_gamma / math.sqrt(2)))
+    return out - k * (C - log_gamma_distrib(0.5 * n, p))
 
 
 def constructL(d, icf):
     constructL.Lparamidx = d
 
     def make_L_col(i):
         nelems = d - i - 1
-        col = torch.cat([torch.zeros(i + 1, dtype=torch.float64), icf[constructL.Lparamidx:(constructL.Lparamidx + nelems)]])
+        col = torch.cat([
+            torch.zeros(i + 1, dtype = torch.float64),
+            icf[constructL.Lparamidx:(constructL.Lparamidx + nelems)]
+        ])
+
         constructL.Lparamidx += nelems
         return col
+
     columns = [make_L_col(i) for i in range(d)]
     return torch.stack(columns, -1)
 
 
 def Qtimesx(Qdiag, L, x):
-    res = Qdiag * x
-    for i in range(L.shape[0]):
-        res = res + L[:, i] * x[i]
-    return res
 
+    f = torch.einsum('rbq,prq->prb', L, x)
```

Comment:
Is this idiomatic?  I would have thought ijk,mik->mij

Reply:
Not, it isn't. I think your option is more convinient